reckoner/tests/test_course.py

```diff
@@ -73,3 +73,39 @@ def test_init_error_fails_min_version_helm(self, configMock, helmClientMock, yam
 
         sysMock.exit.assert_called_once
         return True
+
+
+@mock.patch('reckoner.chart.call', autospec=True)
+@mock.patch('reckoner.repository.Repository', autospec=True)
+@mock.patch('reckoner.course.sys')
+@mock.patch('reckoner.course.yaml', autospec=True)
+@mock.patch('reckoner.course.HelmClient', autospec=True)
+@mock.patch('reckoner.course.Config', autospec=True)
+class TestIntegrationWithChart(unittest.TestCase):
+    def test_failed_pre_install_hooks_fail_chart_installation(self, configMock, helmClientMock, yamlLoadMock, sysMock, repoMock, chartCallMock):
+        """Test that the chart isn't installed when the pre_install hooks return any non-zero responses. This also assures we don't raise python errors with hook errors."""
+        c = configMock()
+        c.helm_args = ['provided args']
+        c.local_development = False
+
+        h = helmClientMock()
+        h.client_version = '0.0.1'
+
+        yamlLoadMock.load.return_value = {
+            'charts': {
+                'first-chart': {
+                    'hooks': {
+                        'pre_install': [
+                            'run a failed command here',
+                        ]
+                    }
+                }
+            }
+        }
+
+        chartCallMock.side_effects = [Response(exitcode=1, command_string='mocked', stderr=' ', stdout=' ')]
```

Comment:
Trying to check this testing.  If I change the exitcode to 0 for this, then the test should fail, correct?

Reply:
hmmm well depends on if everything else in the `.plot()` succeeds. It's not a very effective integration test since there are so many moving pieces between `course.plot()` and `chart.run_hooks().call()`. Have you tried it and the negative testing didn't do what you expected?

Reply:
> I tested the intended functionality, looks good. My only issue (that doesn't necessarily need to be addressed in this PR), is the feedback to the user. It does a great job of outputting the error from the hook, but then there is no indication that the install was skipped because of that. It just says "it failed", which leads the user to believe that the helm install failed. It would be nice to have a "Reason for chart failure" as part of the message there instead of just `Helm upgrade failed on rbac-manager`

This is something I can definitely augment if you'd like. I think it could be something tucked into the class instance of the chart that failed. Maybe another property for "reason failed" so that we can aggregate that at the end of the run.

This addition would be a net new from fixing the bug I originally introduced (correct me if I'm wrong). Since it would be net new we could open an issue for it and pick it up as a future augmentation. I think it would be valuable to be able to see what caused chart failures in general, not only if it's just a pre-install hook.

Reply:
I tried changing the exit code, and  the test passed.  However, if I leave the exit code at 0 and change the expected number of failed charts to 0, the test fails as expected.  

I agree that's a tough testing scenario, but after further playing around I think this is fine for what you are trying to accomplish.

Reply:
Actually looks like this test was broken :D  -- thanks for finding this. It's now fixed. The side_effect wasn't passing the right object. Not sure why still but really this whole part of the code needs some loves to build in better isolation.